### PR TITLE
Move pthread_cancel to native code. NFC.

### PR DESCRIPTION
**Pull request by kleisauke**
Split from PR #13007.

### CODE REVIEW

File: system/lib/pthread/pthread_create.c [hunk 1]
```diff
@@ -244,7 +244,7 @@ void _emscripten_thread_exit(void* result) {
   }
 }
 
-// Mark as `no_sanitize("address"` since emscripten_pthread_exit destroys
+// Mark as `no_sanitize("address")` since emscripten_pthread_exit destroys
```
Comment (sbc100) [inline system/lib/pthread/pthread_create.c:247]: Looks like all the changes to this file are just cleanups.. can you split this out?

Comment (kleisauke) [reply system/lib/pthread/pthread_create.c]: OK, reverted those with commit 72da048d62269fb9b20b2fc23e011594cfeeb81a. I'll make a new PR with these changes soon (perhaps with `[skip ci]` in the commit message, since these are all minor changes?).

File: src/worker.js [hunk 2]
```diff
@@ -280,7 +280,6 @@ self.onmessage = function(e) {
       if (Module['_pthread_self']()) {
         Module['__emscripten_thread_exit'](-1/*PTHREAD_CANCELED*/);
       }
-      postMessage({ 'cmd': 'cancelDone' });
```
Comment (sbc100) [inline src/worker.js:283]: I wonder if the removal on `cancelDone` can be split out into its own change?   It seems like a bugfix its own right as `__emscripten_thread_exit` should always be enough to recycle the thread.    That would be very small bugix PR that would simplify this one.

File: system/lib/pthread/pthread_testcancel.c [hunk 3]
```diff
@@ -8,15 +8,11 @@
...
 
-int _pthread_isduecanceled(struct pthread* pthread_ptr) {
-  return pthread_ptr->cancel != 0;
-}
-
 void __pthread_testcancel() {
   struct pthread* self = pthread_self();
   if (self->canceldisable)
     return;
-  if (_pthread_isduecanceled(self)) {
```
Comment (sbc100) [inline system/lib/pthread/pthread_testcancel.c:19]: Can the removal of _pthread_isduecanceled also be split out into its own PR?

File: system/lib/libc/musl/src/thread/pthread_cond_timedwait.c [hunk 4]
```diff
@@ -187,10 +187,6 @@ int __pthread_cond_timedwait(pthread_cond_t *restrict c, pthread_mutex_t *restri
 		__pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, 0);
 	}
 
-#ifdef __EMSCRIPTEN__
-	pthread_testcancel();
-#endif
-
```
Comment (kleisauke) [inline system/lib/libc/musl/src/thread/pthread_cond_timedwait.c:193]: I removed this since it looked redundant. This was introduced in commit https://github.com/emscripten-core/emscripten/commit/d2bb705cea3a434a936656cd51da889a08be07eb (which was based on musl 1.0.5), but when musl was updated to 1.1.15 (commit https://github.com/emscripten-core/emscripten/commit/31cf2bb1bc3a501a14ea9647f2bc7ad25fb2fce4), it ends up calling `__pthread_testcancel` twice, see:
https://github.com/emscripten-core/emscripten/blob/7e6d511d91ff593c3ec741857fb6d33dbcce59e6/system/lib/libc/musl/src/thread/pthread_cond_timedwait.c#L185-L192

If preferred, I can also create a separate PR for this.

File: src/worker.js (hunk 2 above)
Comment (kleisauke) [reply src/worker.js]: Reverted with commit 3008b16e7fd210aa1cc192bb2f4f77de43eb9b4c, moved to PR #15624.

File: system/lib/pthread/pthread_testcancel.c (hunk 3 above)
Comment (kleisauke) [reply system/lib/pthread/pthread_testcancel.c]: Reverted with commit 8db22b9e14153eae72d20df98aa6d2673807a8f5, moved to PR #15625.

**Comment by kleisauke** [comment]
Rebased, PTAL. This would make reviewing #15604 easier.

File: src/library_pthread.js [hunk 5]
```diff
@@ -408,6 +408,7 @@ var LibraryPThread = {
   },
 
   __emscripten_thread_cleanup: function(thread) {
+    // Called when a thread needs to be cleaned up so it can be reused.
```
Comment (sbc100) [inline src/library_pthread.js:411]: Perhaps mention that its called in exactly two places: "When a detached thread exits and when an attached thread is joined"?

File: system/lib/libc/musl/src/thread/pthread_cancel.c [hunk 6]
```diff
@@ -66,6 +67,7 @@ static void cancel_handler(int sig, siginfo_t *si, void *ctx)
 
 	__syscall(SYS_tkill, self->tid, SIGCANCEL);
 }
+#endif
 
 void __testcancel()
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_cancel.c:72]: If we are including this function, then perhaps we should also be including `system/lib/libc/musl/src/thread/pthread_testcancel.c` rather than our custom `system/lib/pthread/pthread_testcancel.c`?

How about `ifdef`'ing this function out for now and the adding it in as followup PR which includes `system/lib/libc/musl/src/thread/pthread_testcancel.c`?  (i.e. lets keep this PR small and focused.. but we can remove these unused functions).

File: system/lib/libc/musl/src/thread/pthread_cancel.c [hunk 7]
```diff
@@ -14,6 +14,7 @@ long __cancel()
 	return -ECANCELED;
 }
 
+#ifndef __EMSCRIPTEN__
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_cancel.c:17]: I think can move this up to line 5, no?

**Review by sbc100 — APPROVED** [review]
Great!   (lets just ifdef out a little more of the unused code.. but otherwise looks great!)

File: system/lib/libc/musl/src/thread/pthread_cancel.c (hunk 6 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_cancel.c]: Ah you're right, I forgot to adjust the `ifdef`'s when I split PR #15604 (which includes `pthread_testcancel` from musl rather than using a custom one). Fixed with commit f4058e2e8aa4fdda6aed317327d654b898371c2a.

File: system/lib/libc/musl/src/thread/pthread_cancel.c (hunk 7 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_cancel.c]: Yep, fixed with commit f4058e2e8aa4fdda6aed317327d654b898371c2a.

File: src/library_pthread.js (hunk 5 above)
Comment (kleisauke) [reply src/library_pthread.js:411]: Mentioned/elaborated with commit f4058e2e8aa4fdda6aed317327d654b898371c2a.

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Great!

**Comment by sbc100** [comment]
Can you rebase/merge?

**Comment by kleisauke** [comment]
Rebased as requested.
